src/aics-image-viewer/components/TfEditor/index.tsx

```diff
@@ -82,669 +67,313 @@ const TF_GENERATORS: Record<string, (histogram: Histogram) => Lut> = {
-    this.selected = pts[0];
-    this.props.updateChannelLutControlPoints(pts);
-  }
-
-  applyTFGenerator(generator: string): void {
-    const lut = TF_GENERATORS[generator](this.props.channelData.histogram);
-    this.updateControlPointsWithoutColor(lut.controlPoints);
-  }
-
-  createTFGeneratorButton(generator: string, name: string, description: string): React.ReactNode {
-    return (
-      <Tooltip title={description} placement="top">
-        <Button onClick={() => this.applyTFGenerator(generator)}>{name}</Button>
-      </Tooltip>
-    );
-  }
-
-  /////// Public API functions ///////
+  };
 
-  /** Set the pixel data we are manipulating */
-  setData(): void {
-    if (!this.props.channelData) {
-      throw new Error("Transfer Function Editor setData called with no channel data.");
+  const handleChangeColor = (color: ColorResult): void => {
+    lastColorRef.current = colorObjectToArray(color.rgb);
+    if (selectedPointIdx !== null) {
+      const newControlPoints = [...props.controlPoints];
+      newControlPoints[selectedPointIdx].color = lastColorRef.current;
+      setControlPoints(newControlPoints);
     }
-    this.updateScales();
-    this.updateAxis();
-  }
-
-  // Initialize elements and perform the drawing of first drawing
-  ready(): void {
-    // Access the svg dom element
-    this.svg = d3.select(this.svgElement.current);
-    this.width = +this.svg.attr("width") - this.margin.left - this.margin.right;
-    this.height = +this.svg.attr("height") - this.margin.top - this.margin.bottom - 15;
-    this.initializeElements();
-    this.drawChart();
-  }
+  };
 
-  handleCloseColorPicker(): void {
-    this.setState({ colorPickerPosition: null });
+  /** d3-generated svg data string representing both the line between points and the region filled with gradient */
+  const area = useMemo(() => {
+    const areaGenerator = d3
+      .area<ControlPoint>()
+      .x((d) => xScale(d.x))
+      .y0((d) => yScale(d.opacity))
+      .y1(innerHeight)
+      .curve(d3.curveLinear);
+    return areaGenerator(props.controlPoints) ?? undefined;
+  }, [props.controlPoints, xScale, yScale, innerHeight]);
+
+  // The below `useCallback`s are used as "ref callbacks" - passed as the `ref` prop of SVG elements in order to render
+  // these elements' content using D3. They are called when the ref'd component mounts and unmounts, and whenever their
+  // identity changes (i.e. whenever their dependencies change).
+
+  const xAxisRef = useCallback(
+    (el: SVGGElement) => {
+      const ticks = xScale.ticks(TFEDITOR_NUM_TICKS);
+      ticks[ticks.length - 1] = xScale.domain()[1];
+      d3.select(el).call(d3.axisBottom(xScale).tickValues(ticks));
+    },
+    [xScale]
+  );
+
+  const yAxisRef = useCallback(
+    (el: SVGGElement) => d3.select(el).call(d3.axisLeft(yScale).ticks(TFEDITOR_NUM_TICKS)),
+    [yScale]
+  );
+
+  const histogramRef = useCallback(
+    (el: SVGGElement) => {
+      if (el === null) {
+        return;
```

Comment:
when is it null and when it becomes non-null does this code get re-run?

Reply:
The behavior of ref callbacks is that the function will be called with `null` whenever the component unmounts - in this case, whenever volume settings are closed. When the component remounts, they get called with an element again.